packages/xstate-react/src/useMachine.ts

```diff
@@ -163,44 +166,47 @@ export function useMachine<
   >([]);
 
   useLayoutEffect(() => {
-    service.onTransition((currentState) => {
-      if (currentState.changed) {
-        setState(currentState);
-      }
-
-      if (currentState.actions.length) {
-        const reactEffectActions = currentState.actions.filter(
-          (action): action is ReactActionObject<TContext, TEvent> => {
-            return (
-              typeof action.exec === 'function' &&
-              '__effect' in (action as ReactActionObject<TContext, TEvent>).exec
-            );
-          }
-        );
-
-        const [effectActions, layoutEffectActions] = partition(
-          reactEffectActions,
-          (action): action is ReactActionObject<TContext, TEvent> => {
-            return action.exec.__effect === ReactEffectType.Effect;
-          }
-        );
-
-        effectActionsRef.current.push(
-          ...effectActions.map<ActionStateTuple<TContext, TEvent>>(
-            (effectAction) => [effectAction, currentState]
-          )
-        );
-
-        layoutEffectActionsRef.current.push(
-          ...layoutEffectActions.map<ActionStateTuple<TContext, TEvent>>(
-            (layoutEffectAction) => [layoutEffectAction, currentState]
-          )
-        );
-      }
-    });
-
-    // if service.state has not changed React should just bail out from this update
-    setState(service.state);
+    service
+      .onTransition((currentState) => {
+        // Only change the current state if:
+        // - the incoming state is the initial state (since it might have new actors)
+        // - OR the incoming state actually changed
+        if (currentState.changed || currentState.changed === undefined) {
+          setState(currentState);
+        }
+
+        if (currentState.actions.length) {
+          const reactEffectActions = currentState.actions.filter(
+            (action): action is ReactActionObject<TContext, TEvent> => {
+              return (
+                typeof action.exec === 'function' &&
+                '__effect' in
+                  (action as ReactActionObject<TContext, TEvent>).exec
+              );
+            }
+          );
+
+          const [effectActions, layoutEffectActions] = partition(
+            reactEffectActions,
+            (action): action is ReactActionObject<TContext, TEvent> => {
+              return action.exec.__effect === ReactEffectType.Effect;
+            }
+          );
+
+          effectActionsRef.current.push(
+            ...effectActions.map<ActionStateTuple<TContext, TEvent>>(
+              (effectAction) => [effectAction, currentState]
+            )
+          );
+
+          layoutEffectActionsRef.current.push(
+            ...layoutEffectActions.map<ActionStateTuple<TContext, TEvent>>(
+              (layoutEffectAction) => [layoutEffectAction, currentState]
+            )
+          );
+        }
+      })
+      .start(rehydratedState ? State.create(rehydratedState) : undefined);
```

Comment:
we have already checked `rehydratedState` when initializing the state, could we just pass `state` here? 🤔 or do we have to recreate `rehydratedState` again?

Reply:
The logic is a little bit different (and causes tests to fail) - we need to have the logic that the initial state is `undefined` so that the service doesn't try to "reuse" the initial state.

Probably can be refactored in a later PR.